solidity/contracts/Rewards.sol

```diff
@@ -0,0 +1,611 @@
+/// and the quota of how many keeps must be created in an interval
+/// for the full reward for that interval to be paid out.
+///
+/// The amount of KEEP to be distributed is determined by funding the contract,
+/// and additional KEEP can be added at any time.
+/// The reward contract is funded with `approveAndCall` with no extra data,
+/// but it also collects any KEEP mistakenly sent to it in any other way.
+///
+/// An interval is defined by the timestamps [startOf, endOf);
+/// a keep created at the time `startOf(i)` belongs to interval `i`
+/// and one created at `endOf(i)` belongs to `i+1`.
+///
+/// When an interval is over,
+/// it will be allocated a percentage of the remaining unallocated rewards
+/// based on its weight,
+/// and adjusted by the number of keeps created in the interval
+/// if the quota is not met.
+/// The adjustment for not meeting the keep quota is a percentage
+/// that equals the percentage of the quota that was met;
+/// if the number of keeps created is 80% of the quota
+/// then 80% of the base reward will be allocated for the interval.
+///
+/// Any unallocated rewards will stay in the unallocated rewards pool,
+/// to be allocated for future intervals.
+/// Intervals past the initially defined schedule have a weight of 100%,
+/// meaning that all remaining unallocated rewards
+/// will be allocated to the interval.
+///
+/// Keeps of the appropriate type can receive rewards
+/// once the interval they were created in is over,
+/// and the keep has closed happily.
+/// There is no time limit to receiving rewards,
+/// nor is there need to wait for all keeps from the interval to close.
+/// Calling `receiveReward` automatically allocates the rewards
+/// for the interval the specified keep was created in
+/// and all previous intervals.
+///
+/// If a keep is terminated,
+/// that fact can be reported to the reward contract.
+/// Reporting a terminated keep returns its allocated reward
+/// to the pool of unallocated rewards.
+///
+/// A concrete implementation of the abstract rewards contract
+/// must specify functions for accessing information about keeps
+/// and paying out rewards.
+/// For the purpose of rewards,
+/// Random Beacon signing groups count as "keeps"
+/// and the beacon operator contract acts as the "factory".
+contract Rewards {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    IERC20 public token;
+
+    // Total number of keep tokens to distribute.
+    uint256 public totalRewards;
+    // Rewards that haven't been allocated to finished intervals
+    uint256 public unallocatedRewards;
+    // Rewards that have been paid out;
+    // `token.balanceOf(address(this))` should always equal
+    // `totalRewards.sub(paidOutRewards)`
+    uint256 public paidOutRewards;
+
+    // Length of one interval in seconds (timestamp diff).
+    uint256 public termLength;
+    // Timestamp of first interval beginning.
+    uint256 public firstIntervalStart;
```

Comment:
Yes, that's correct. I'll clarify it in the documentation. If we need to extend rewards further into the past without using the backporting, a further parameter can be added to prevent allocation of intervals until enough time has passed from the creation of the contract.